modules/dag-l1/src/main/scala/org/tessellation/dag/l1/storage/TransactionStorage.scala

```diff
@@ -0,0 +1,111 @@
+
+import io.chrisdavenport.mapref.MapRef
+
+trait TransactionStorage[F[_]] {
+  def isParentAccepted(tx: Transaction): F[Boolean]
+
+  def getLastAcceptedTransactionRef(address: Address): F[TransactionReference]
+
+  def acceptTransaction(hashedTx: Hashed[Transaction]): F[Unit]
+
+  def putTransactions(transactions: Set[Signed[Transaction]]): F[Unit]
+
+  def pullTransactions(): F[Option[NonEmptyList[Signed[Transaction]]]]
+}
+
+object TransactionStorage {
+
+  def make[F[_]: Sync]: F[TransactionStorage[F]] =
+    for {
+      lastAccepted <- MapRef.ofConcurrentHashMap[F, Address, TransactionReference]()
+      waitingTransactions <- Ref.of[F, Map[Address, NonEmptySet[Signed[Transaction]]]](Map.empty)
+      transactionStorage = make(lastAccepted, waitingTransactions)
+    } yield transactionStorage
+
+  def make[F[_]: Sync](
+    lastAccepted: MapRef[F, Address, Option[TransactionReference]],
+    waitingTransactions: Ref[F, Map[Address, NonEmptySet[Signed[Transaction]]]]
+  ): TransactionStorage[F] =
+    new TransactionStorage[F] {
+
+      def isParentAccepted(transaction: Transaction): F[Boolean] =
+        (transaction.parent == TransactionReference.empty)
+          .pure[F]
+          .ifM(
+            true.pure[F],
+            lastAccepted(transaction.source).get.map(_.contains(transaction.parent))
+          )
+
+      def getLastAcceptedTransactionRef(source: Address): F[TransactionReference] =
+        lastAccepted(source).get.map(_.getOrElse(TransactionReference.empty))
+
+      // Either transactions should be accepted in the correct order (handled outside of this scope),
+      // or maybe we should check if we can update here? (current < new) then transactions accepted in wrong order
+      // will still end up in consistent state - though it feels kind of like a side effect that should be out of scope of storage
+      def acceptTransaction(hashedTx: Hashed[Transaction]): F[Unit] = {
+        val source = hashedTx.signed.value.source
+        val reference = TransactionReference(hashedTx.hash, hashedTx.signed.value.ordinal)
+
+        lastAccepted(source).set(reference.some)
+      }
+
+      def putTransactions(transactions: Set[Signed[Transaction]]): F[Unit] =
+        waitingTransactions.modify { current =>
+          val updatedWaiting =
+            transactions.foldRight(current) {
+              case (tx, waiting) =>
+                val updatedAddressTxs = waiting
+                  .get(tx.value.source)
+                  .map(txs => NonEmptySet.of(tx, txs.toNonEmptyList.toList: _*))
+                  .getOrElse(NonEmptySet.one(tx))
+
+                waiting.updated(tx.value.source, updatedAddressTxs)
+            }
+
+          (updatedWaiting, ())
+        }
+
+      // I would like to pull inside of a modify so I'm prefetching the whole lastAccepted (which could theoretically
+      // change during the pull which increases invalid block creation probability - though I'm not sure that's something worry about it)
```

Comment:
we had this issue before. pull should be an atomic change and we shouldn't split it into two parts - prefetch and then modify. if you need to do so, I'd use a semaphore to lock the access

Reply:
Actually after thinking about it a bit more this scenario seems ok. I'm prefetching the `lastAcceptedTransaction` and then atomically pulling from `waitingTransactions` based on `lastAcceptedTransaction`. Even if in the meantime new transaction is being accepted which unblocks transactions from waiting during pull is happening - it's ok, simply the transactions will be pulled on the next go. We may have to use semaphores when we will have redownload where the state could actually be rolled back. Until we have that we don't need the semaphore I think.